app/schema/schema.go

```diff
@@ -139,15 +141,84 @@ func Action_GoCodegen(args *cli.Context) error {
 	outputDir := args.Path("output")
 	pkgName := args.String("package")
 
-	if generator != "go-gengo" {
+	switch generator {
+	case "go-gengo":
+		if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+			return err
+		}
+		a := gengo.AdjunctCfg{}
+		gengo.Generate(outputDir, pkgName, ts, &a)
+	case "go-bindnode":
+		if err := generateGoBindnode(schemaFilePath, outputDir, pkgName, &ts); err != nil {
+			return err
+		}
+	default:
 		return fmt.Errorf("unsupported generator: %s", generator)
 	}
 
+	return nil
+}
+
+func generateGoBindnode(schemaFilePath, outputDir, pkgName string, ts *schema.TypeSystem) error {
 	if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
 		return err
 	}
-	a := gengo.AdjunctCfg{}
-	gengo.Generate(outputDir, pkgName, ts, &a)
 
-	return nil
+	// generate types.go
+	f, err := os.Create(fmt.Sprintf("%s/types.go", outputDir))
+	if err != nil {
+		return err
+	}
+
+	if _, err := fmt.Fprintf(f, "package %s\n\n", pkgName); err != nil {
+		return err
+	}
+
+	if err := bindnode.ProduceGoTypes(f, ts); err != nil {
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	relPath, err := filepath.Rel(outputDir, schemaFilePath)
```

Comment:
My intuition is that basically noone will want to use both gengo and bindnode codegen at the same time. Users will want to choose one or the other. At most, they might switch from one to the other, if they find they chose the wrong one given their tradeoffs/priorities.

Reply:
Copy plus opt-out seems fine.

May be worth mentioning here: @rvagg has at some point produced a tool which would let you have your IPLD Schema interspersed in a markdown file (or multiple markdown files) by just plucking out all code blocks tagged "`ipldsch`".  I don't think we have that capability in our golang code today.  But I have to say, I thought the idea was rather cool and ideally would like to keep that possibility floating around and relatively accessible.  It's very nice for making specifications and documentation converge.

Reply:
As far as I can tell, opting out is not possible: If the idea of "opt-out" is to go:embed the original .ipldsch file (instead of the copied one), then this requires a go:embed directive with a relative path (like go:embed ../some.ipldsch). This is not supported by go:embed. Further, I don't see reasons for the opt out feature: The output directory is fully dedicated to (and owned by) code generation, and the user should have no preferences as to what goes there. Having choices only complicates their mental model.

Reply:
I don't have particularly strong feelings here. It seems like "copy the files around with go:generate" is the pattern https://github.com/golang/go/issues/46056, although perhaps @mvdan or @warpfork have other ideas.

As long as the user can reasonably have a single place to edit their schema no matter which/how many codegens they're using then SGTM.